File: src/utils/encoding.ts
/**
 * PDF string encoding utilities.
 */

export function latin1Encode(s: string): Uint8Array {
  const result = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) {
    result[i] = s.charCodeAt(i) & 0xff;
  }
  return result;
}

export function latin1Decode(bytes: Uint8Array): string {
  let s = '';
  for (let i = 0; i < bytes.length; i++) {
    s += String.fromCharCode(bytes[i]);
  }
  return s;
}

export function utf16BEEncode(s: string): Uint8Array {
  // BOM (0xFE, 0xFF) + 2 bytes per code unit
  const codeUnits: number[] = [];
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    codeUnits.push(code);
  }
  const result = new Uint8Array(2 + codeUnits.length * 2);
  result[0] = 0xfe;
  result[1] = 0xff;
  for (let i = 0; i < codeUnits.length; i++) {
    result[2 + i * 2] = (codeUnits[i] >>> 8) & 0xff;
    result[2 + i * 2 + 1] = codeUnits[i] & 0xff;
  }
  return result;
}

export function utf16BEDecode(bytes: Uint8Array): string {
  let offset = 0;
  // Skip BOM if present
  if (bytes.length >= 2 && bytes[0] === 0xfe && bytes[1] === 0xff) {
    offset = 2;
  }
  let s = '';
  for (let i = offset; i + 1 < bytes.length; i += 2) {
    const code = (bytes[i] << 8) | bytes[i + 1];
    s += String.fromCharCode(code);
  }
  return s;
}

/**
 * PDFDocEncoding to Unicode mapping for bytes 0x80-0xFF that differ from Latin-1.
 * Unmapped entries use 0xFFFD (replacement character).
 */
const PDF_DOC_ENCODING_MAP: Record<number, number> = {
  0x80: 0x2022, // BULLET
  0x81: 0x2020, // DAGGER
  0x82: 0x2021, // DOUBLE DAGGER
  0x83: 0x2026, // HORIZONTAL ELLIPSIS
  0x84: 0x2014, // EM DASH
  0x85: 0x2013, // EN DASH
  0x86: 0x0192, // LATIN SMALL LETTER F WITH HOOK
  0x87: 0x2044, // FRACTION SLASH
  0x88: 0x2039, // SINGLE LEFT-POINTING ANGLE QUOTATION MARK
  0x89: 0x203a, // SINGLE RIGHT-POINTING ANGLE QUOTATION MARK
  0x8a: 0x2212, // MINUS SIGN
  0x8b: 0x2030, // PER MILLE SIGN
  0x8c: 0x201e, // DOUBLE LOW-9 QUOTATION MARK
  0x8d: 0x201c, // LEFT DOUBLE QUOTATION MARK
  0x8e: 0x201d, // RIGHT DOUBLE QUOTATION MARK
  0x8f: 0x2018, // LEFT SINGLE QUOTATION MARK
  0x90: 0x2019, // RIGHT SINGLE QUOTATION MARK
  0x91: 0x201a, // SINGLE LOW-9 QUOTATION MARK
  0x92: 0x2122, // TRADE MARK SIGN
  0x93: 0xfb01, // LATIN SMALL LIGATURE FI
  0x94: 0xfb02, // LATIN SMALL LIGATURE FL
  0x95: 0x0141, // LATIN CAPITAL LETTER L WITH STROKE
  0x96: 0x0152, // LATIN CAPITAL LIGATURE OE
  0x97: 0x0160, // LATIN CAPITAL LETTER S WITH CARON
  0x98: 0x0178, // LATIN CAPITAL LETTER Y WITH DIAERESIS
  0x99: 0x017d, // LATIN CAPITAL LETTER Z WITH CARON
  0x9a: 0x0131, // LATIN SMALL LETTER DOTLESS I
  0x9b: 0x0142, // LATIN SMALL LETTER L WITH STROKE
  0x9c: 0x0153, // LATIN SMALL LIGATURE OE
  0x9d: 0x0161, // LATIN SMALL LETTER S WITH CARON
  0x9e: 0x017e, // LATIN SMALL LETTER Z WITH CARON
  0xa0: 0x20ac, // EURO SIGN
};

export function pdfDocEncodingDecode(bytes: Uint8Array): string {
  let s = '';
  for (let i = 0; i < bytes.length; i++) {
    const b = bytes[i];
    if (b < 0x80) {
      s += String.fromCharCode(b);
    } else if (b in PDF_DOC_ENCODING_MAP) {
      s += String.fromCharCode(PDF_DOC_ENCODING_MAP[b]);
    } else {
      // For bytes 0x80-0xFF not in the map, use Latin-1 identity mapping
      s += String.fromCharCode(b);
    }
  }
  return s;
}

export function isAscii(s: string): boolean {
  for (let i = 0; i < s.length; i++) {
    if (s.charCodeAt(i) > 0x7e) return false;
  }
  return true;
}

export function textToUtf16BEHex(s: string): string {
  const bytes = utf16BEEncode(s);
  let hex = '';
  for (let i = 0; i < bytes.length; i++) {
    hex += bytes[i].toString(16).padStart(2, '0').toUpperCase();
  }
  return hex;
}

File: src/utils/encoding.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const EncodingError = error{
    InvalidUtf8,
    InvalidUtf16,
    Unmappable,
};

/// PDFDocEncoding table: maps byte values 0x80..0x9F to Unicode codepoints.
/// Values outside this range map 1:1 with Unicode (same as Latin-1 for 0xA0..0xFF).
const pdf_doc_encoding_extra = [32]u21{
    0x2022, // 0x80 BULLET
    0x2020, // 0x81 DAGGER
    0x2021, // 0x82 DOUBLE DAGGER
    0x2026, // 0x83 HORIZONTAL ELLIPSIS
    0x2014, // 0x84 EM DASH
    0x2013, // 0x85 EN DASH
    0x0192, // 0x86 LATIN SMALL LETTER F WITH HOOK
    0x2044, // 0x87 FRACTION SLASH
    0x2039, // 0x88 SINGLE LEFT-POINTING ANGLE QUOTATION MARK
    0x203A, // 0x89 SINGLE RIGHT-POINTING ANGLE QUOTATION MARK
    0x2212, // 0x8A MINUS SIGN
    0x2030, // 0x8B PER MILLE SIGN
    0x201E, // 0x8C DOUBLE LOW-9 QUOTATION MARK
    0x201C, // 0x8D LEFT DOUBLE QUOTATION MARK
    0x201D, // 0x8E RIGHT DOUBLE QUOTATION MARK
    0x2018, // 0x8F LEFT SINGLE QUOTATION MARK
    0x2019, // 0x90 RIGHT SINGLE QUOTATION MARK
    0x201A, // 0x91 SINGLE LOW-9 QUOTATION MARK
    0x2122, // 0x92 TRADE MARK SIGN
    0xFB01, // 0x93 LATIN SMALL LIGATURE FI
    0xFB02, // 0x94 LATIN SMALL LIGATURE FL
    0x0141, // 0x95 LATIN CAPITAL LETTER L WITH STROKE
    0x0152, // 0x96 LATIN CAPITAL LIGATURE OE
    0x0160, // 0x97 LATIN CAPITAL LETTER S WITH CARON
    0x0178, // 0x98 LATIN CAPITAL LETTER Y WITH DIAERESIS
    0x017D, // 0x99 LATIN CAPITAL LETTER Z WITH CARON
    0x0131, // 0x9A LATIN SMALL LETTER DOTLESS I
    0x0142, // 0x9B LATIN SMALL LETTER L WITH STROKE
    0x0153, // 0x9C LATIN SMALL LIGATURE OE
    0x0161, // 0x9D LATIN SMALL LETTER S WITH CARON
    0x017E, // 0x9E LATIN SMALL LETTER Z WITH CARON
    0xFFFD, // 0x9F UNDEFINED -> REPLACEMENT CHARACTER
};

/// Convert Latin-1 encoded data to UTF-8.
pub fn latin1ToUtf8(allocator: Allocator, data: []const u8) Allocator.Error![]u8 {
    var result: std.ArrayListUnmanaged(u8) = .empty;
    errdefer result.deinit(allocator);

    for (data) |byte| {
        if (byte < 0x80) {
            try result.append(allocator, byte);
        } else {
            const codepoint: u21 = byte;
            var buf: [4]u8 = undefined;
            const utf8_len = std.unicode.utf8Encode(codepoint, &buf) catch unreachable;
            try result.appendSlice(allocator, buf[0..utf8_len]);
        }
    }

    return try result.toOwnedSlice(allocator);
}

/// Convert UTF-8 encoded data to Latin-1. Returns error for unmappable characters.
pub fn utf8ToLatin1(allocator: Allocator, data: []const u8) (Allocator.Error || EncodingError)![]u8 {
    var result: std.ArrayListUnmanaged(u8) = .empty;
    errdefer result.deinit(allocator);

    var view = std.unicode.Utf8View.initUnchecked(data);
    var iter = view.iterator();

    while (iter.nextCodepoint()) |cp| {
        if (cp <= 0xFF) {
            try result.append(allocator, @intCast(cp));
        } else {
            return EncodingError.Unmappable;
        }
    }

    return try result.toOwnedSlice(allocator);
}

/// Encode a UTF-8 string as UTF-16BE with BOM (0xFEFF prefix).
pub fn utf16beEncode(allocator: Allocator, text: []const u8) (Allocator.Error || EncodingError)![]u8 {
    var result: std.ArrayListUnmanaged(u8) = .empty;
    errdefer result.deinit(allocator);

    // Write BOM
    try result.appendSlice(allocator, &[_]u8{ 0xFE, 0xFF });

    var view = std.unicode.Utf8View.initUnchecked(text);
    var iter = view.iterator();

    while (iter.nextCodepoint()) |cp| {
        if (cp < 0x10000) {
            const val: u16 = @intCast(cp);
            try result.appendSlice(allocator, &std.mem.toBytes(std.mem.nativeToBig(u16, val)));
        } else {
            const adjusted = cp - 0x10000;
            const high: u16 = @intCast(0xD800 + (adjusted >> 10));
            const low: u16 = @intCast(0xDC00 + (adjusted & 0x3FF));
            try result.appendSlice(allocator, &std.mem.toBytes(std.mem.nativeToBig(u16, high)));
            try result.appendSlice(allocator, &std.mem.toBytes(std.mem.nativeToBig(u16, low)));
        }
    }

    return try result.toOwnedSlice(allocator);
}

/// Decode UTF-16BE data (with or without BOM) to UTF-8.
pub fn utf16beDecode(allocator: Allocator, data: []const u8) (Allocator.Error || EncodingError)![]u8 {
    var result: std.ArrayListUnmanaged(u8) = .empty;
    errdefer result.deinit(allocator);

    if (data.len % 2 != 0) return EncodingError.InvalidUtf16;

    var offset: usize = 0;

    // Skip BOM if present
    if (data.len >= 2 and data[0] == 0xFE and data[1] == 0xFF) {
        offset = 2;
    }

    while (offset + 1 < data.len) {
        const unit: u16 = @as(u16, data[offset]) << 8 | data[offset + 1];
        offset += 2;

        var codepoint: u21 = undefined;

        if (unit >= 0xD800 and unit <= 0xDBFF) {
            if (offset + 1 >= data.len) return EncodingError.InvalidUtf16;
            const low: u16 = @as(u16, data[offset]) << 8 | data[offset + 1];
            offset += 2;
            if (low < 0xDC00 or low > 0xDFFF) return EncodingError.InvalidUtf16;
            codepoint = @as(u21, unit - 0xD800) << 10 | @as(u21, low - 0xDC00);
            codepoint += 0x10000;
        } else if (unit >= 0xDC00 and unit <= 0xDFFF) {
            return EncodingError.InvalidUtf16;
        } else {
            codepoint = unit;
        }

        var buf: [4]u8 = undefined;
        const utf8_len = std.unicode.utf8Encode(codepoint, &buf) catch return EncodingError.InvalidUtf16;
        try result.appendSlice(allocator, buf[0..utf8_len]);
    }

    return try result.toOwnedSlice(allocator);
}

/// Convert PDFDocEncoding data to UTF-8.
/// PDFDocEncoding is similar to Latin-1 but differs in the 0x80..0x9F range.
pub fn pdfDocEncodingToUtf8(allocator: Allocator, data: []const u8) Allocator.Error![]u8 {
    var result: std.ArrayListUnmanaged(u8) = .empty;
    errdefer result.deinit(allocator);

    for (data) |byte| {
        const codepoint: u21 = if (byte >= 0x80 and byte <= 0x9F)
            pdf_doc_encoding_extra[byte - 0x80]
        else
            byte;

        if (codepoint < 0x80) {
            try result.append(allocator, @intCast(codepoint));
        } else {
            var buf: [4]u8 = undefined;
            const utf8_len = std.unicode.utf8Encode(codepoint, &buf) catch unreachable;
            try result.appendSlice(allocator, buf[0..utf8_len]);
        }
    }

    return try result.toOwnedSlice(allocator);
}

test "latin1ToUtf8: ASCII passthrough" {
    const result = try latin1ToUtf8(std.testing.allocator, "Hello");
    defer std.testing.allocator.free(result);
    try std.testing.expectEqualSlices(u8, "Hello", result);
}

test "latin1ToUtf8: high bytes" {
    const input = &[_]u8{ 0x48, 0xE9, 0x6C, 0x6C, 0x6F };
    const result = try latin1ToUtf8(std.testing.allocator, input);
    defer std.testing.allocator.free(result);
    try std.testing.expectEqualSlices(u8, "H\xc3\xa9llo", result);
}

test "utf8ToLatin1: roundtrip" {
    const original = &[_]u8{ 0x48, 0xE9, 0x6C, 0x6C, 0x6F };
    const utf8 = try latin1ToUtf8(std.testing.allocator, original);
    defer std.testing.allocator.free(utf8);
    const back = try utf8ToLatin1(std.testing.allocator, utf8);
    defer std.testing.allocator.free(back);
    try std.testing.expectEqualSlices(u8, original, back);
}

test "utf8ToLatin1: unmappable" {
    const result = utf8ToLatin1(std.testing.allocator, "\xe2\x80\xa2");
    try std.testing.expectError(EncodingError.Unmappable, result);
}

test "utf16beEncode and decode roundtrip" {
    const input = "Hello, World!";
    const encoded = try utf16beEncode(std.testing.allocator, input);
    defer std.testing.allocator.free(encoded);

    try std.testing.expectEqual(@as(u8, 0xFE), encoded[0]);
    try std.testing.expectEqual(@as(u8, 0xFF), encoded[1]);

    const decoded = try utf16beDecode(std.testing.allocator, encoded);
    defer std.testing.allocator.free(decoded);
    try std.testing.expectEqualSlices(u8, input, decoded);
}

test "utf16beEncode: supplementary characters" {
    const input = "\xF0\x9F\x98\x80";
    const encoded = try utf16beEncode(std.testing.allocator, input);
    defer std.testing.allocator.free(encoded);

    const decoded = try utf16beDecode(std.testing.allocator, encoded);
    defer std.testing.allocator.free(decoded);
    try std.testing.expectEqualSlices(u8, input, decoded);
}

test "utf16beDecode: invalid odd length" {
    const result = utf16beDecode(std.testing.allocator, &[_]u8{ 0x00, 0x41, 0x00 });
    try std.testing.expectError(EncodingError.InvalidUtf16, result);
}

test "pdfDocEncodingToUtf8: ASCII passthrough" {
    const result = try pdfDocEncodingToUtf8(std.testing.allocator, "test");
    defer std.testing.allocator.free(result);
    try std.testing.expectEqualSlices(u8, "test", result);
}

test "pdfDocEncodingToUtf8: special range 0x80" {
    const input = &[_]u8{0x80};
    const result = try pdfDocEncodingToUtf8(std.testing.allocator, input);
    defer std.testing.allocator.free(result);
    try std.testing.expectEqualSlices(u8, "\xe2\x80\xa2", result);
}

test "pdfDocEncodingToUtf8: latin-1 range preserved" {
    const input = &[_]u8{0xE9};
    const result = try pdfDocEncodingToUtf8(std.testing.allocator, input);
    defer std.testing.allocator.free(result);
    try std.testing.expectEqualSlices(u8, "\xc3\xa9", result);
}
